presto-accumulo/src/main/java/com/facebook/presto/accumulo/index/ColumnCardinalityCache.java

```diff
@@ -0,0 +1,398 @@
+
+            // Collect all exact Accumulo Ranges, i.e. single value entries vs. a full scan
+            Collection<Range> exactRanges =
+                    colValues.stream().filter(x -> isExact(x)).collect(Collectors.toList());
+            LOG.debug("Column values contain %s exact ranges of %s", exactRanges.size(),
+                    colValues.size());
+
+            // Sum the cardinalities for the exact-value Ranges
+            // This is where the reach-out to Accumulo occurs for all Ranges that have not
+            // previously been fetched
+            long sum = 0;
+            for (Long e : cache.getAll(exactRanges).values()) {
+                sum += e;
+            }
+
+            // If these collection sizes are not equal,
+            // then there is at least one non-exact range
+            if (exactRanges.size() != colValues.size()) {
+                // for each range in the column value
+                for (Range r : colValues) {
+                    // if this range is not exact
+                    if (!isExact(r)) {
+                        // Then get the value for this range using the single-value cache lookup
+                        long val = cache.get(r);
+
+                        // add our value to the cache and our sum
+                        cache.put(r, val);
+                        sum += val;
+                    }
+                }
+            }
+
+            LOG.debug("Cache stats : size=%s, %s", cache.size(), cache.stats());
+            return sum;
+        }
+
+        /**
+         * Gets a Boolean value indicating if the given Range is an exact value
+         *
+         * @param r Range to check
+         * @return True if exact, false otherwise
+         */
+        private boolean isExact(Range r)
+        {
+            return !r.isInfiniteStartKey() && !r.isInfiniteStopKey()
+                    && r.getStartKey().followingKey(PartialKey.ROW).equals(r.getEndKey());
+        }
+
+        /**
+         * Creates a new cache for the given column
+         *
+         * @param schema Schema name
+         * @param table Table name
+         * @param family Accumulo column family for the column
+         * @param qualifier Accumulo qualifier for the column
+         * @return A fresh LoadingCache
+         */
+        private LoadingCache<Range, Long> newCache(String schema, String table, String family,
+                String qualifier)
+        {
+            LOG.debug("Created new cache for %s.%s, column %s:%s, size %d expiry %d", schema, table,
+                    family, qualifier, size, expireSeconds);
+            return CacheBuilder.newBuilder().maximumSize(size)
+                    .expireAfterWrite(expireSeconds, TimeUnit.SECONDS)
+                    .build(new CardinalityCacheLoader(schema, table, family, qualifier));
+        }
+    }
+
+    /**
+     * Internal class for loading the cardinality from Accumulo
+     */
+    private class CardinalityCacheLoader
```

Comment:
Please annotate with `@ThreadSafe` to indicate it fulfills `CacheLoader` requirements for thread-safety.